src/views/CHConfigEditorHooks.ts

```diff
@@ -35,3 +35,59 @@ export const useMigrateV3Config = (options: DataSourceSettings<CHConfig>, onOpti
     }
   }, [v3ServerField, v3TimeoutField, options, onOptionsChange]);
 };
+
+/**
+ * Handles saving HTTP headers to Grafana config.
+ * 
+ * All header keys go to the unsecure config.
+ * If the header is marked as secure, its value goes to the
+ * secure json config where it is hidden.
+ */
+export const onHttpHeadersChange = (headers: CHHttpHeader[], options: DataSourceSettings<CHConfig, CHSecureConfig>, onOptionsChange: (opts: DataSourceSettings<CHConfig, CHSecureConfig>) => void) => {
+  const httpHeaders: CHHttpHeader[] = [];
+  const secureHttpHeaderKeys: KeyValue<boolean> = {};
+  const secureHttpHeaderValues: KeyValue<string> = {};
+
+  for (let header of headers) {
+    if (!header.name) {
+      continue;
+    }
+
+    if (header.secure) {
+      const key = `secureHttpHeaders.${header.name}`;
+      secureHttpHeaderKeys[key] = true;
+
+      if (header.value) {
+        secureHttpHeaderValues[key] = header.value;
+        header.value = '';
+      }
+    }
+
+    httpHeaders.push(header);
+  }
+
+  const currentSecureJsonFields: KeyValue<boolean> = { ...options.secureJsonFields };
+  for (let key in currentSecureJsonFields) {
+    if (!secureHttpHeaderKeys[key] && key.startsWith('secureHttpHeaders.')) {
```

Comment:
I don't know a better way but I wish we could have this synced somehow between go/TS so if it gets changed in one place it throws a compiler error or something.

Reply:
I tried something with the TS [template literal types](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html), but there's no way to really enforce it considering that other secure keys might be added to the config object.